src/components/input/SelectNext.tsx

```diff
@@ -0,0 +1,182 @@
+import classnames from 'classnames';
+import type { ComponentChildren } from 'preact';
+import {
+  useCallback,
+  useContext,
+  useId,
+  useLayoutEffect,
+  useRef,
+  useState,
+} from 'preact/hooks';
+
+import { useArrowKeyNavigation } from '../../hooks/use-arrow-key-navigation';
+import { useClickAway } from '../../hooks/use-click-away';
+import { useKeyPress } from '../../hooks/use-key-press';
+import { MenuCollapseIcon, MenuExpandIcon } from '../icons';
+import Button from './Button';
+import SelectContext from './SelectContext';
+
+export type SelectProps<T> = {
+  selected: T;
+  onChange: (newValue: T) => void;
+  label: ComponentChildren;
+  children: ComponentChildren;
+  disabled?: boolean;
+};
+
+function SelectMain<T>({
+  label,
+  selected,
+  onChange,
+  children,
+  disabled,
+}: SelectProps<T>) {
+  const [isDropdownOpen, setIsDropdownOpen] = useState(false);
+  const closeDropdown = useCallback(() => setIsDropdownOpen(false), []);
+  const wrapperRef = useRef<HTMLDivElement | null>(null);
+  const listboxId = useId();
+  const buttonRef = useRef<HTMLButtonElement | null>(null);
+  const buttonId = useId();
+  const [shouldDropUp, setShouldDropUp] = useState(false);
+
+  const selectValue = useCallback(
+    (newValue: unknown) => {
+      onChange(newValue as T);
+      closeDropdown();
+    },
+    [closeDropdown, onChange],
+  );
+
+  // When clicking away or pressing `Esc`, close the dropdown
+  useClickAway(wrapperRef, closeDropdown);
+  useKeyPress(['Escape'], closeDropdown);
+
+  // Vertical arrow key for options in the dropdown
+  useArrowKeyNavigation(wrapperRef, { horizontal: false });
+
+  // Focus button after closing dropdown
+  useLayoutEffect(() => {
+    if (!isDropdownOpen) {
+      buttonRef.current!.focus();
+    }
+  }, [isDropdownOpen]);
+
+  if (isDropdownOpen) {
+    const viewportHeight = window.innerHeight;
+    const { top: buttonDistanceToTop, height } =
+      buttonRef.current!.getBoundingClientRect();
+    const buttonDistanceToBottom =
+      viewportHeight - (buttonDistanceToTop + height);
+
+    // The listbox should drop up only if there's not enough space below it for
+    // the listbox max height, and there's also more space above
+    setShouldDropUp(
+      // 320px is tailwind's h-80. e should try not to couple with this
+      buttonDistanceToBottom < 320 &&
+        buttonDistanceToTop > buttonDistanceToBottom,
+    );
+  }
+
+  return (
+    <div className="relative" ref={wrapperRef}>
+      <Button
+        id={buttonId}
+        variant="custom"
+        classes={classnames(
+          'w-full flex border',
+          'bg-grey-0 disabled:bg-grey-1 disabled:text-grey-6',
+        )}
+        expanded={isDropdownOpen}
+        pressed={isDropdownOpen}
+        disabled={disabled}
+        aria-haspopup="listbox"
```

Comment:
For me VoiceOver is reading the button as a toggle button, rather than a toggle button that controls a listbox. This might be because the listbox is not rendered, rather than merely hidden, until the dropdown is opened.

Reply:
I can confirm orca has the same behavior, identifying the button as a toggle.

If I change the listbox to be hidden instead of unmounted, then it identifies it properly as a listbox handler.